### PR TITLE
add stream api for websocket message pushing and code refactor

### PR DESCRIPTION
**Pull request by huojiao2006**

### CODE REVIEW

File: api/notification.proto [hunk 1]
```diff
@@ -405,7 +405,27 @@ message ValidateEmailServiceResponse{
...
+
+message StreamResData {
+	UserMessage userMsg = 1;
+}
+
+message UserMessage {
+	string user_id = 1;
+	string service = 2;
+	string message_type = 3;
+	MessageDetail msgDetail= 4;
```
Comment (chilianyi) [inline api/notification.proto]: Use same mode。

msg_detail

google.protobuf.StringValue

File: pkg/pubsub/i_pubsub.go [hunk 2]
```diff
@@ -0,0 +1,16 @@
+// Copyright 2019 The OpenPitrix Authors. All rights reserved.
```
Comment (chilianyi) [inline pkg/pubsub/i_pubsub.go]: Not to use abbreviation will be better，i means a lot.

File: pkg/pubsub/redis_pubsub.go [hunk 3]
```diff
@@ -2,71 +2,80 @@
...
 
-		go getWsMessagesByRredis(pubsub, c)
+		go rPubSub.getWsMessagesByRredis(pubsub, userMsgChan)
 	}
-	return c
+	return userMsgChan
 }
 
-func getWsMessagesByRredis(ps *r.PubSub, c chan wstypes.UserMessage) {
+func (rPubSub *RedisPubSub) getWsMessagesByRredis(ps *redis.PubSub, userMsgChan chan models.UserMessage) {
```
Comment (chilianyi) [inline pkg/pubsub/redis_pubsub.go]: getWsMessagesByRedis

File: pkg/services/notification/handler.go [hunk 4]
```diff
@@ -513,3 +513,16 @@ func (s *Server) ValidateEmailService(ctx context.Context, req *pb.ServiceConfig
 		IsSucc: pbutil.ToProtoBool(true),
 	}, nil
 }
+
+func (s *Server) GetStream(req *pb.StreamReqData, res pb.Notification_GetStreamServer) error {
+	for {
+		outmsg := <-s.controller.websocketMsgChan
```
Comment (chilianyi) [inline pkg/services/notification/handler.go]: outMsg

Comment (chilianyi) [reply pkg/services/notification/handler.go]: Different user should use different stream.

File: api/notification.proto (hunk 1 above)
Comment (huojiao2006) [reply api/notification.proto]: update

File: pkg/pubsub/i_pubsub.go (hunk 2 above)
Comment (huojiao2006) [reply pkg/pubsub/i_pubsub.go]: update

File: api/notification.proto [hunk 5]
```diff
@@ -568,5 +590,8 @@ service notification {
 		};
 	}
 
+	rpc GetDataFromGrpcStream (StreamReqData) returns (stream StreamResData){
```
Comment (chilianyi) [inline api/notification.proto]: CreateNotificationChannel will be better.

File: pkg/services/notification/handler.go [hunk 6]
```diff
@@ -513,3 +520,21 @@ func (s *Server) ValidateEmailService(ctx context.Context, req *pb.ServiceConfig
 		IsSucc: pbutil.ToProtoBool(true),
 	}, nil
 }
+
+func (s *Server) GetDataFromGrpcStream(req *pb.StreamReqData, res pb.Notification_GetDataFromGrpcStreamServer) error {
+	service := req.GetService().GetValue()
+	for {
+		outMsg := <-s.controller.websocketMsgChan
```
Comment (chilianyi) [inline pkg/services/notification/handler.go]: If op and ks use same notification, then with the same websocketMsgChan. 

if op get ks's notification, may abandon it.

File: pkg/services/notification/resource_control/notification.go [hunk 7]
```diff
@@ -230,47 +227,60 @@ func processsAddressInfo4AddressMap(ctx context.Context, notification *models.No
...
-			UserId:        userId,
-			Service:       service,
-			MessageType:   messageType,
-			MessageDetail: msgDetail,
-		}
-		err = websocket.PushWsMessage(context.Background(), config.GetInstance().PubSub.Type, global.GetInstance().GetPubSubClient(), &userMsg)
+		err = ipubsub.Publish(userMsgStr)
 		if err != nil {
-			logger.Errorf(ctx, "Push user [%s] websocket message id [%s] failed: %+v", userId, task.TaskId, err)
+			logger.Errorf(ctx, "Push websocket message Directive[%s] to pubsub failed: %+v", jsonutil.ToString(task.Directive), err)
```
Comment (chilianyi) [inline pkg/services/notification/resource_control/notification.go]: Directive -> directive

File: pkg/client/websocket/ws_manager.go [hunk 8]
```diff
@@ -166,25 +162,38 @@ func (wsm *wsManager) HandleWsTask() func(w http.ResponseWriter, r *http.Request
...
+	}
+
+	res, err := client.GetDataFromGrpcStream(ctx, reqstreamData)
+	if err != nil {
+		logger.Errorf(nil, "failed to get msgs by grpc stream,err=%+v", err)
+	}
 
-		err := wsMessageRedis.PublishWsMessage(ctx, &rClient, userMsg)
+	for {
+		userWsMsgStreamData, err := res.Recv()
```
Comment (chilianyi) [inline pkg/client/websocket/ws_manager.go]: Need check err here.

File: pkg/client/websocket/ws_manager.go [hunk 9]
```diff
@@ -166,25 +162,38 @@ func (wsm *wsManager) HandleWsTask() func(w http.ResponseWriter, r *http.Request
...
-		wsMessageEtcd := new(ws_message_etcd.WsMessageEtcd)
-		eClient := (*psClient).(wstypes.EtcdClient)
-		err := wsMessageEtcd.PushWsMessage(ctx, &eClient, userMsg)
-		if err != nil {
-			return err
-		}
-	} else if pubsubType == "redis" {
-		wsMessageRedis := new(ws_message_redis.WsMessageRedis)
-		rClient := (*psClient).(wstypes.RedisClient)
+func (wsm *wsManager) ReceiverMsgFromNf() chan models.UserMessage {
```
Comment (chilianyi) [inline pkg/client/websocket/ws_manager.go]: ReceiverMsgFromNf -> ReceiveMsg

File: pkg/services/notification/controller.go [hunk 10]
```diff
@@ -23,33 +26,61 @@ import (
...
+	var wsMsgStrChan chan string
+
+	if queueType == constants.QueueTypeRedis {
+		//queue setting
+		redisQueue := qredis.RedisQueue{}
+		iqueue = &redisQueue
+
+		//pubsub setting
+		if config.GetInstance().Websocket.Enable {
+			redisPubSub := qredis.RedisPubSub{}
```
Comment (chilianyi) [inline pkg/services/notification/controller.go]: Should not care about queueType: redis or etcd. 

ipubsub = queue.NewPubSub(queueType) will be better.

File: api/notification.proto (hunk 5 above)
Comment (huojiao2006) [reply api/notification.proto]: update

File: pkg/services/notification/resource_control/notification.go (hunk 7 above)
Comment (huojiao2006) [reply pkg/services/notification/resource_control/notification.go]: done

File: pkg/client/websocket/ws_manager.go (hunk 9 above)
Comment (huojiao2006) [reply pkg/client/websocket/ws_manager.go]: done

File: api/notification.proto (hunk 5 above)
Comment (huojiao2006) [reply api/notification.proto]: fix

File: pkg/services/notification/resource_control/notification.go (hunk 7 above)
Comment (huojiao2006) [reply pkg/services/notification/resource_control/notification.go]: fix

File: pkg/client/websocket/ws_manager.go (hunk 9 above)
Comment (huojiao2006) [reply pkg/client/websocket/ws_manager.go]: fix

File: pkg/services/notification/controller.go (hunk 10 above)
Comment (huojiao2006) [reply pkg/services/notification/controller.go]: fix

File: pkg/client/websocket/ws_manager.go (hunk 8 above)
Comment (huojiao2006) [reply pkg/client/websocket/ws_manager.go]: fix

File: api/notification.proto [hunk 11]
```diff
@@ -568,5 +590,8 @@ service notification {
 		};
 	}
 
+	rpc CreateNotificationChannel (StreamReqData) returns (stream StreamResData){
```
Comment (chilianyi) [inline api/notification.proto]: res-> resp will be better

File: pkg/config/config.go [hunk 12]
```diff
@@ -75,8 +67,7 @@ type Config struct {
 	}
 
 	Websocket struct {
-		//ServiceMessageTypes string `default:"op/nf,op/event,ks/nf,ks/event"`
-		ServiceMessageTypes string `default:"none"`
+		Service string `default:"op,ks"`
```
Comment (chilianyi) [inline pkg/config/config.go]: none will be better. If ks or op use websocket, then need to set the service in env or config or api.

**Review by chilianyi — APPROVED** [review]

**Pull request merged**